README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Tango

Chaccour and Brew

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
<!-- badges: end -->

Tango. It’s an intricate and complex dance, with high levels of
coordination and timing between different roles (dancers, bandoleon, and
other musicians). It’s also the tentative name of this project: an
intricate, complex attempt at [“the
dance”](https://medium.com/@tomaspueyo/coronavirus-the-hammer-and-the-dance-be9337092b56),
ie the long-term effort to control COVID-19 until there’s a vaccine.

In scientific lingo, we might call this project something like “Guding
post-lockdown COVID-19 confinement policies through an assessment of
spatial-temporal heterogeneity in population vulnerability and
receptivity”.

## The concepts

### A. Receptivity and vulnerability

We borrow two concepts from the world of [malaria
research](https://www.who.int/malaria/mpac/mpac-april2019-session7-report-erg-malariogenic-potential.pdf?ua=1):

1.  Receptivity: how “well-suited” an area is for high-risk cases

2.  Vulnerability: the frequency of influx of infected individuals or
    groups

Understanding the concept through examples:

  - An rural village with a young, healty population would be:
      - Low receptivity (not many people who are likely to develop a
        severe case)
      - Low vulnerability  
  - A coastal town with a high prevalence of elderly retirees:
      - High receptivity (many people predisposed to a severe case due
        to age)
      - High vulnerability (high risk of disease “importation” due to
        tourist movement)
  - A commuter town with a close proximity to an urban center would be:
      - Low receptivity (mostly working-age people below age of
        highest-risk)
      - High vulnerability (lots of movement to and from other areas)
  - An isolated mountain town with an elderly population might be:
      - High receptivity (many people predisposed to severe illness due
        to age)
      - Low vulnerability (very little in/out flow)

### B. Susceptibility and spatial-temporal infection risk

Neither receptivity nor vulnerability can be understood in a vacuum.
Both require incorporating the concepts of susceptibility and
spatial-temporal infection risk

1.  Susceptibility: how much of the population remains uninfected and is
    therefore susceptible to becoming infected and infectious

2.  Spatial-temporal risk: how much disease activity is there nearby
    (understanding “nearby” to be a function of actual flows of human
    mobility, not simple linear distance) at the time in question

### Integrating the concepts: an index

We propose a risk “index”, which takes into account all four of the
aforementioned concepts:

  - A1. Receptivity
  - A2. Vulnerability
  - B1. Susceptibility
  - B2. Spatial-temporal risk

Once constructed, this index serves to determine the level of
“loosening” of social distancing measures so as to (i) minimize loss
of life and health, (ii) maximize social and economic activity, and
(iii) implement the *correct* degree of disease control measures (ie,
robust enough to prevent contagion, but not overly robust so as to
restrict human activity or lead to poor compliance).

## Work-flow

  - A1. Receptivity: We can assess this right now with publicly
    available census data (see below)
  - A2. Vulnerability: We can assess this right now, roughly, by a
    simple flow matrix based on population density (ie, we assume flows
    are between unpopulated areas and more densely populated areas)
  - B1. Susceptibility: We need to carry out mass serology, sampling a
    random but significantly large percentage of the at-risk population
    in every municipality
  - B2. Spatial-temporal risk: We need a constant flow of
    epidemiological data (similar to what the Ministry publishes now,
    but at the municipality level)

# Receptivity

First, we’ll prepare data for analysis.

``` r
# Load the package for this project
library(tango)
# Load other useful packages
library(knitr)
library(tidyr)
library(dplyr)
library(sp)
library(ggplot2)
library(RColorBrewer)

# Get census age data
census <- tango::census
# Get municipios spatial data
municipios <- tango::municipios
```

We’ll define a function for getting “receptivity” (ie, age-based risk)
for each municipality. In this first iteration, we’ll just set it as the
percentage of people who are above age `n`

``` r
define_receptivity <- function(data, n){
  data %>%
    mutate(receptive = edad >= n) %>%
    summarise(pop_receptive = sum(total[receptive], na.rm = TRUE),
              total_pop = sum(total, na.rm = TRUE)) %>%
    ungroup %>%
    mutate(p_receptive = pop_receptive / total_pop * 100)
}
```

We then define receptivity for each municipio in Spain, with an age
cut-off of, for example, 80 years:

``` r
risks <- census %>%
  group_by(municipio, id) %>%
  define_receptivity(n = 80) %>%
  arrange(desc(p_receptive))
```

Let’s take a peak at the most “receptive” municipalities (ie, those
whose populations are most pre-disposed to severe
cases):

``` r
risks %>% head %>% kable
```

| municipio            | id    | pop\_receptive | total\_pop | p\_receptive |
| :------------------- | :---- | -------------: | ---------: | -----------: |
| Riba de Escalote, La | 42157 |             10 |         18 |     55.55556 |
| Olmeda de la Cuesta  | 16140 |             18 |         37 |     48.64865 |
| Ciruelos del Pinar   | 19089 |             16 |         34 |     47.05882 |
| Valdemadera          | 26161 |              4 |          9 |     44.44444 |
| Tormellas            | 05244 |             28 |         65 |     43.07692 |
| Tórtoles             | 05246 |             30 |         73 |     41.09589 |

The below shows the distribution of percentage of people 80 or older by
municipality.

``` r
ggplot(data = risks,
       aes(x = p_receptive)) +
  geom_density(fill = 'darkorange', alpha = 0.6) +
  theme_simple() +
  geom_text(data = tibble(x = c(12, 40),
                          y = c(0.01, 0.01),
                          label = c('Very many\nlow-receptivity municipalities\n(can have more relaxed measures)',
                                      'Very few especially\nhigh-receptivity municipalities\n(need tighter measures)')),
            aes(x = x, y = y, label = label)) +
  labs(title = 'Distribution of receptivity by municipality',
       subtitle = '"Receptivity"= % of inhabitants 80 or older')
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" />

Let’s think of this another way. If we created an arbitrary cut-off for
relaxing certain measures at, for example, a receptivity of 25% or
lower, we would see:

``` r
x = risks %>%
  group_by(status = ifelse(p_receptive <= 25, 'Relax', 'No-relax')) %>%
  summarise(municipalities = n(),
            population = sum(total_pop)) %>%
  mutate(percentage = round(population / sum(population) * 100, digits = 2))
names(x) <- Hmisc::capitalize(names(x))
x %>% kable
```

| Status   | Municipalities | Population | Percentage |
| :------- | -------------: | ---------: | ---------: |
| No-relax |            206 |      18738 |       0.04 |
| Relax    |           7910 |   46789993 |      99.96 |

Let’s map receptivity (again, using the 80 years cut-off example).

``` r
map <- municipios
map@data <- left_join(map@data, risks, by = 'id')
mycolours <- brewer.pal(8, "YlOrRd")
spplot(map, 'p_receptive', par.settings = list(axis.line = list(col ="transparent")), main = "% of population 80 or over by municipality", cuts = 7, col ="transparent", col.regions = mycolours)
```

<img src="man/figures/README-unnamed-chunk-8-1.png" width="100%" />

We can vary a bit. Let’s do 70 years…

``` r
risks <- census %>%
  group_by(municipio, id) %>%
  define_receptivity(n = 70) %>%
  arrange(desc(p_receptive))
map <- municipios
map@data <- left_join(map@data, risks, by = 'id')
mycolours <- brewer.pal(8, "YlOrRd")
spplot(map, 'p_receptive', par.settings = list(axis.line = list(col ="transparent")), main = "% of population 70 or over by municipality", cuts = 7, col ="transparent", col.regions = mycolours)
```

<img src="man/figures/README-unnamed-chunk-9-1.png" width="100%" />

and 60 years…

``` r
risks <- census %>%
  group_by(municipio, id) %>%
  define_receptivity(n = 60) %>%
  arrange(desc(p_receptive))
map <- municipios
map@data <- left_join(map@data, risks, by = 'id')
mycolours <- brewer.pal(8, "YlOrRd")
spplot(map, 'p_receptive', par.settings = list(axis.line = list(col ="transparent")), main = "% of population 60 or over by municipality", cuts = 7, col ="transparent", col.regions = mycolours)
```

<img src="man/figures/README-unnamed-chunk-10-1.png" width="100%" />

Good. Now we’ve identified especially “receptive” populations (ie, those
with an age structure that puts them at risk). Task A1 (1 of 4) done.

Time to do the other three tasks:

  - A2. Vulnerability
  - B1. Susceptibility
  - B2. Spatial-temporal risk

# Vulnerability

For vulnerability, we want to estimate how much movement there is
between each municipality and other areas. We could do this with raw
data from, for example, mobile networks. But until we have those data,
we’ll use a more simple metric: how many nearby population centers of
mobile people are there.

Let’s call a population center any place with a population of \>5,000
people. And let’s say nearby = 20km. And let’s say “mobile people” are
those between ages 18 and 65.

Here are the population centers:

``` r
library(geosphere)
# Get cities populations
cities_sp <- census %>%
  filter(edad >= 18,
         edad <= 65) %>%
  group_by(id, municipio) %>%
  summarise(population = sum(total, na.rm = TRUE)) %>%
  ungroup %>%
  filter(population >= 5000)
# Get locations
x <- municipios
x@data <- left_join(x@data, cities_sp)
x <- x[!is.na(x@data$population),]

plot(municipios)
plot(x, add = T, col = 'red')
```

<img src="man/figures/README-unnamed-chunk-11-1.png" width="100%" />

``` r
cities_sp <- x
# Define matrix of distances
distances <- geosphere::distm(x = coordinates(municipios),
                              y = coordinates(cities_sp),
                              fun = distHaversine)
distances <- t(distances)
# distances <- rgeos::gDistance(spgeom1 = municipios,
#                               cities_sp,
#                               byid = T)
out <- rep(NA, ncol(distances))
for(j in 1:ncol(distances)){
  this_municipality <- municipios@data$NAMEUNIT[j] 
  distance_values <- distances[,j]
  keeps <- which(distance_values < 20000)
  n <- length(keeps)
  out[j] <- n
}
# Now, for every municipality, we have a vulnerability score
map <- municipios
map@data$vulnerability <- sqrt(out)
```

Having calculated vulnerability score, let’s visualize

``` r
mycolours <- brewer.pal(8, "YlOrRd")
spplot(map, 'vulnerability', par.settings = list(axis.line = list(col ="transparent")), main = "Vulnerability score", cuts = 7, col ="transparent", col.regions = mycolours)
```

<img src="man/figures/README-unnamed-chunk-13-1.png" width="100%" />

Now we can *cross* vulnerability and receptivity, to identify those
areas which are most at risk demographically (ie, old) and by virtue of
potential mobility (ie, proximity to population centers). Here it is:

``` r
risks <- census %>%
  group_by(municipio, id) %>%
  define_receptivity(n = 80) %>%
  arrange(desc(p_receptive))
map@data <- left_join(map@data, risks, by = 'id')
# Create index based on both vulnerability and receptivity
map@data$index <- map@data$vulnerability * map@data$p_receptive

spplot(map, 'index', par.settings = list(axis.line = list(col ="transparent")), main = "Index score\n(using vulnerability and receptivity)", cuts = 7, col ="transparent", col.regions = mycolours)
```

<img src="man/figures/README-unnamed-chunk-14-1.png" width="100%" />

Great. We have some concept of vulnerability and vulnerability, and we
build a basic index on both. We’re two steps in the right direction.

But the index can be improved. Being close to other populations (ie,
being “vulnerable”) only matters if those other populations are
infectious. Which brings us to our next steps…

# Next steps

What’s next? We need to assess (1) susceptibility and (1)
spatial-temporal risk. Urgently

For (1), we need funds and a team to start preparations for mass
serology. Sampling strategy and timing will be fundamental to getting
actionable data. This will almost certainly be the most expensive, but
also most important, part of this work.

For (2), we need the Ministry of Health to release municipality-level
data as soon as possible. This should be granular (ie, a dataset with
one row per case, with variables including municipality, date of symptom
onset, age, sex, place of diagnosis).

# Technical set-up

This analysis is set-up as an R package. One can install it in one of
two ways:

1.  Clone from <https://github.com/joebrew/tango> and then build package
    from source (sequentially walking through the code in
    `data-raw/update.R` and then running `update.sh` from the command
    line).

2.  In an R session, install the package directly from github:
    `devtools::install_github('joebrew/tango')`.


NEWS.md
# tango 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(tango)

test_check("tango")


dev/01_start.R
# Building a Prod-Ready, Robust Shiny Application.
# 
# README: each step of the dev files is optional, and you don't have to 
# fill every dev scripts before getting started. 
# 01_start.R should be filled at start. 
# 02_dev.R should be used to keep track of your development during the project.
# 03_deploy.R should be used once you need to deploy your app.
# 
# 
########################################
#### CURRENT FILE: ON START SCRIPT #####
########################################

## Fill the DESCRIPTION ----
## Add meta data about your application
golem::fill_desc(
  pkg_name = "tango", # The Name of the package containing the App 
  pkg_title = "Tango", # The Title of the package containing the App 
  pkg_description = "Tools for identifying populations at risk in the COVID-19 outbreak", # The Description of the package containing the App 
  author_first_name = "Joe", # Your First Name
  author_last_name = "Brew", # Your Last Name
  author_email = "joe@databrew.cc", # Your Email
  repo_url = "https://github.com/joebrew/tango" # The URL of the GitHub Repo (optional) 
)     

## Set {golem} options ----
golem::set_golem_options()

## Create Common Files ----
## See ?usethis for more information
usethis::use_mit_license( name = "Joe Brew" )  # You can set another license here
usethis::use_readme_rmd( open = FALSE )
usethis::use_code_of_conduct()
usethis::use_lifecycle_badge( "Experimental" )
usethis::use_news_md( open = FALSE )

## Use git ----
usethis::use_git()

## Init Testing Infrastructure ----
## Create a template for tests
golem::use_recommended_tests()

## Use Recommended Packages ----
golem::use_recommended_deps()

## Favicon ----
# If you want to change the favicon (default is golem's one)
golem::remove_favicon()
golem::use_favicon() # path = "path/to/ico". Can be an online file. 

## Add helper functions ----
golem::use_utils_ui()
golem::use_utils_server()

# You're now set! ----

# go to dev/02_dev.R
rstudioapi::navigateToFile( "dev/02_dev.R" )



data-raw/update_data.R
library(dplyr)
library(readr)
library(usethis)
library(raster)
library(sp)
library(rgdal)
library(readxl)
library(rmapshaper)
library(sf)

esp0 <- getData(country = 'ESP', level = 0)
usethis::use_data(esp0, overwrite = TRUE)

esp1 <- getData(country = 'ESP', level = 1)
usethis::use_data(esp1, overwrite = TRUE)

# Municipal codes
codes <- read_excel('20codmun.xlsx', skip = 1)

## Get Spanish census data
# Population by age and municipality
# https://www.ine.es/jaxi/Tabla.htm?path=/t20/e244/avance/p02/l0/&file=1mun00.px&L=0
download_url <- 'https://www.ine.es/jaxi/files/_px/es/csv_bdsc/t20/e244/avance/p02/l0/1mun00.csv_bdsc?nocab=1'
download_file <- '1mun00.csv'
if(!download_file %in% dir()){
  download.file(url = download_url,
                destfile = download_file)
}
# Read in
census <- read_delim(download_file, delim = ';')
# Clean up names
names(census) <- c('municipio',
                   'sexo',
                   'edad',
                   'periodio',
                   'total')
census$periodio <- NULL
# Remove unnecessary values
census <- census %>%
  filter(!municipio %in% 'TOTAL NACIONAL',
         !sexo %in% 'Ambos sexos',
         !edad %in% 'Total')
census$total <- ifelse(census$total == '..',
                       0,
                       as.numeric(as.character(gsub('.', '', 
                                                    census$total, fixed = T))))
census$edad <- gsub('años|año', '', census$edad)
census$edad <- gsub(' y más', '', census$edad)
census$edad <- as.numeric(as.character(census$edad))

census$id <- unlist(lapply(strsplit(census$municipio, ' '),
                           function(x){paste0(x[1], collapse = ' ')}))
census$id <- trimws(census$id)
# census$id <- as.numeric(census$id)
census$municipio <- unlist(lapply(strsplit(census$municipio, ' '),
                                  function(x){paste0(x[2:length(x)], collapse = ' ')}))
census$municipio <- trimws(census$municipio)
usethis::use_data(census, overwrite = TRUE)

# Get Spain at municipal level
# Must download teh following as 'lineas_limite.zip'
# http://centrodedescargas.cnig.es/CentroDescargas/catalogo.do?Serie=CAANE
# and then unzip
# unzip('lineas_limite.zip')
municipios <- readOGR('recintos_municipales_inspire_peninbal_etrs89',
                      'recintos_municipales_inspire_peninbal_etrs89', encoding="UTF-8")
# municipios <- st_read('recintos_municipales_inspire_peninbal_etrs89/recintos_municipales_inspire_peninbal_etrs89.shp')
# municipios <- read_sf('recintos_municipales_inspire_peninbal_etrs89/recintos_municipales_inspire_peninbal_etrs89.shp')
library(rgeos)
regions_df <- municipios@data
x = gSimplify(municipios, tol = 0.05, topologyPreserve = TRUE)

municipios <- sp::SpatialPolygonsDataFrame(x, regions_df)


# Get an ID in municipios
census_ids <- sort(unique(census$id))
census_ids <- sample(census_ids, length(census_ids))
ids <- substr(municipios$INSPIREID, 20, 24)

municipios$id <- ids
usethis::use_data(municipios, overwrite = TRUE)

# World cities
cities <- read_csv('worldcities.csv')
cities <- cities %>% filter(country == 'Spain')
cities <- citie %>% dplyr::select(city, lat, lng, population)
cities_sp <- cities %>% dplyr::mutate(x = lng, y = lat)
coordinates(cities_sp) <- ~x+y
proj4string(cities_sp) <- proj4string(municipios)
usethis::use_data(cities, overwrite = T)
usethis::use_data(cities_sp, overwrite = T)


R/app.R

##################################################
# UI
##################################################
#' @import shiny
#' @import shinydashboard
#' @import leaflet
#' @import shiny
#' @import ggplot2
#' @import shinyMobile
#' @import leaflet
#' @import RColorBrewer
#' @import sp
#' @import DT
#' @import dplyr
mobile_app_ui <- function(request) {
  options(scipen = '999')
  
  tagList(
    mobile_golem_add_external_resources(),
    
    f7Page(
      init = f7Init(
        skin = 'ios', #  c("ios", "md", "auto", "aurora"),
        theme = 'light', #c("dark", "light"),
        filled = TRUE
      ),
      title = "Tango",
      f7SingleLayout(
        navbar = f7Navbar(
          title = "Tango",
          hairline = TRUE,
          shadow = TRUE
        ),
        toolbar = f7Toolbar(
          position = "bottom",
          f7Link(label = "Carlos Chaccour", src = "https://www.isglobal.org/person?p_p_id=viewpersona_WAR_intranetportlet&p_p_lifecycle=0&p_p_col_id=column-1&p_p_col_count=1&_viewpersona_WAR_intranetportlet_struts_action=%2Fview%2FpersonaView&_viewpersona_WAR_intranetportlet_personaId=7304", external = TRUE),
          f7Link(label = "Joe Brew", src = "https://databrew.cc", external = TRUE)
          
        ),
        # main content
        f7Shadow(
          intensity = 10,
          hover = TRUE,
          f7Card(
            sliderInput('age_at_risk',
                        'Age of "risk"',
                        min = 0, max = 100,
                        value = c(80, 100)),
            # sliderInput('vulnerability_age',
            #             'Age of "vulnerability"',
            #             min = 0, max = 100,
            #             value = c(20, 50)),
            checkboxInput('overlay_deaths',
                     'Overlay observed deaths at CA level?',
                      value = FALSE),
            selectInput('show',
                        'Show on map:',
                        choices = c('Risk'#,
                                    # 'Vulnerability'
                                    )),
                                    # 'Product of receptivity and vulnerability (combined index)')),
            br(),
            f7Button('generate_map',
                         'Generate map and tables'),
            br(),
            helpText('Note: generating the map may take up to 30 seconds. Be patient'),
            br(),
            f7Button('clear_map',
                         'Clear map'),
            br(),
            leafletOutput('the_plot', height = '500px'),
            DT::dataTableOutput('the_table'),
            br(),
            sliderInput('cut_off',
                        'Cut-off threshold',
                        min = 0, max = 100, value = 10),
            DT::dataTableOutput('summary_table'),
            
            br(),
            f7Button('generate_threshold_map',
                     'Generate THRESHOLD map'),
            br(),
            helpText('Note: generating the map may take up to 30 seconds. Be patient'),
            leafletOutput('threshold_plot', height = '500px'),
            
            height = 500
          )
        )
      )
    )
  )
}

#' Add external Resources to the Application
#' 
#' This function is internally used to add external 
#' resources inside the Shiny application. 
#' 
#' @import shiny
#' @importFrom golem add_resource_path activate_js favicon bundle_resources
#' @noRd
mobile_golem_add_external_resources <- function(){
  # addResourcePath(
  #   'www', system.file('app/www', package = 'covid19')
  # )
  
  share <- list(
    title = "Tango",
    url = "http://joebrew.net",
    image = "http://www.databrew.cc/images/blog/covid2.png",
    description = "Joe Brew / Tango",
    twitter_user = "joethebrew"
  )
  
  tags$head(
    
    # Facebook OpenGraph tags
    tags$meta(property = "og:title", content = share$title),
    tags$meta(property = "og:type", content = "website"),
    tags$meta(property = "og:url", content = share$url),
    tags$meta(property = "og:image", content = share$image),
    tags$meta(property = "og:description", content = share$description),
    
    # Twitter summary cards
    tags$meta(name = "twitter:card", content = "summary"),
    tags$meta(name = "twitter:site", content = paste0("@", share$twitter_user)),
    tags$meta(name = "twitter:creator", content = paste0("@", share$twitter_user)),
    tags$meta(name = "twitter:title", content = share$title),
    tags$meta(name = "twitter:description", content = share$description),
    tags$meta(name = "twitter:image", content = share$image),
    
    # golem::activate_js(),
    # golem::favicon(),
    # Add here all the external resources
    # Google analytics script
    # includeHTML(system.file('app/www/google-analytics-mini.html', package = 'covid19')),
    # includeScript(system.file('app/www/script.js', package = 'covid19')),
    # includeScript(system.file('app/www/mobile.js', package = 'covid19')),
    # includeScript('inst/app/www/script.js'),
    
    # includeScript('www/google-analytics.js'),
    # If you have a custom.css in the inst/app/www
    tags$link(rel="stylesheet", type="text/css", href="www/custom.css")
    # tags$link(rel="stylesheet", type="text/css", href="www/custom.css")
  )
}

##################################################
# SERVER
##################################################
#' @import shiny
#' @import leaflet
mobile_app_server <- function(input, output, session) {
  
  define_risk <- function(data, n1, n2){
    data %>%
      mutate(at_risk = edad >= n1 & edad <= n2) %>%
      summarise(pop_at_risk = sum(total[at_risk], na.rm = TRUE),
                total_pop = sum(total, na.rm = TRUE)) %>%
      ungroup %>%
      mutate(p_at_risk = pop_at_risk / total_pop * 100)
  }
  # define_vulnerability <- function(data, n1, n2){
  #   data %>%
  #     mutate(vulnerable = edad >= n1 & edad <= n2) %>%
  #     summarise(pop_vulnerable = sum(total[vulnerable], na.rm = TRUE),
  #               total_pop = sum(total, na.rm = TRUE)) %>%
  #     ungroup %>%
  #     mutate(p_vulnerable = pop_vulnerable / total_pop * 100)
  # }
  
  data_list <- reactiveValues(data = data.frame())
  
  output$the_plot <- renderLeaflet({
    leaflet() %>%
      addTiles() %>%
      addPolygons(data = esp0)
  })
  
  output$threshold_plot <- renderLeaflet({
    leaflet() %>%
      addTiles() %>%
      addPolygons(data = esp0)
  })
  
  output$the_table <- DT::renderDataTable({
    out <- data_list$data
    if(!is.null(out)){
      if(nrow(out) > 0){
        out <- out  %>% dplyr::select(id, municipio,
                                      pop_at_risk,
                                      # pop_vulnerable,
                                      p_at_risk,
                                      # p_vulnerable,
                                      total_pop)
      }
    }
    out 
  })
  
  output$summary_table <- DT::renderDataTable({
    cut_off <- input$cut_off
    out <- data_list$data
    pd <- NULL
    if(!is.null(out)){
      if(nrow(out) > 0){
        pd <- out %>%
          mutate(`Category` = ifelse(p_at_risk >= cut_off,
                                     'Above threshold',
                                     'Below threshold')) %>%
          group_by(Category) %>%
          summarise(`Number of municipalities` = n(),
                    `Number of people in those municipalities` = sum(total_pop, na.rm = TRUE))
      }
    }
    pd 
  })
  
  observeEvent(input$generate_map, {
    
    # Get receptivity
    ns <- input$age_at_risk
    message('check1')
    risks <- census %>%
      group_by(municipio, id) %>%
      define_risk(n1 = ns[1],
                         n2 = ns[2]) %>%
      arrange(desc(p_at_risk))
    message('check2')

    map <- municipios
    map@data <- left_join(map@data, risks, by = 'id')

    map <- municipios
    map@data <- left_join(map@data, risks, by = 'id')
    show <- input$show
    map@data$var <- map@data$p_at_risk
    data_list$data <- map@data
    
    
    p_popup <- paste0(map@data$NAMEUNIT, ': Percent ', show, ': ',  round(map@data$var, digits = 2))

    pal_fun <- colorQuantile("YlOrRd", NULL, n = 8)
    leafletProxy("the_plot") %>%
      clearControls() %>%
      clearShapes() %>% 
        addPolygons(
          # highlightOptions = highlightOptions(stroke = 4, weight = 2),
          data = map,
          stroke = FALSE, # remove polygon borders
          fillColor = ~pal_fun(var), # set fill color with function from above and value
          fillOpacity = 0.8, smoothFactor = 0.5, # make it nicer
          label = p_popup)
      
  })
  
  observeEvent(input$generate_threshold_map, {
    
    # Get receptivity
    ns <- input$age_at_risk
    risks <- census %>%
      group_by(municipio, id) %>%
      define_risk(n1 = ns[1],
                  n2 = ns[2]) %>%
      arrange(desc(p_at_risk))

    map <- municipios
    map@data <- left_join(map@data, risks, by = 'id')
    
    map <- municipios
    map@data <- left_join(map@data, risks, by = 'id')
    show <- input$show
    map@data$var <- map@data$p_at_risk
    threshold <- input$cut_off
    map@data$var <- ifelse(map@data$var >= threshold, 'Above threshold',
                           'Below threshold')

    p_popup <- paste0(map@data$NAMEUNIT, ': Category ', show, ': ',  map@data$var)
    
    pal_fun <- colorFactor("YlOrRd", map@data$var, reverse = TRUE)
    leafletProxy("threshold_plot") %>%
      clearControls() %>%
      clearShapes() %>% 
      addPolygons(
        # highlightOptions = highlightOptions(stroke = 4, weight = 2),
        data = map,
        stroke = FALSE, # remove polygon borders
        fillColor = ~pal_fun(var), # set fill color with function from above and value
        fillOpacity = 0.8, smoothFactor = 0.5, # make it nicer
        label = p_popup)
    
  })
  
  observeEvent(input$clear_map, {
    leafletProxy("the_plot") %>%
      clearControls() %>%
      clearShapes() 
    
  })
}

app <- function(){
  # Detect the system. If on AWS, don't launch browswer
  is_aws <- grepl('aws', tolower(Sys.info()['release']))
  shinyApp(ui = mobile_app_ui,
           server = mobile_app_server,
           options = list('launch.browswer' = !is_aws))
}